### PR TITLE
[android][expo-document-picker] Migrate to new modules API

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
Continue migration to new modules API

# How
Followed the usual migration steps

# Test Plan
Tested in the bare expo app. Everything working as before.

### CODE REVIEW

File: packages/expo-document-picker/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,9 @@
 
 ### 🎉 New features
 
+- Migrated to Expo Modules API. ([#20336](https://github.com/expo/expo/pull/20336)
+  by [@alanhughes](https://github.com/alanjhughes))
```
Comment (lukmccall) [inline packages/expo-document-picker/CHANGELOG.md:10]:
```suggestion
- Migrated to Expo Modules API. ([#20336](https://github.com/expo/expo/pull/20336) by [@alanhughes](https://github.com/alanjhughes))
```

File: packages/expo-document-picker/android/src/main/java/expo/modules/documentpicker/DocumentPickerModule.kt [hunk 2]
```diff
@@ -5,116 +5,93 @@ import android.content.Context
...
+          val result = Bundle().apply {
+            putString("type", "success")
+            putString("uri", details.uri)
+            putString("name", details.name)
+            details.mimeType?.let { mimeType ->
+              putString("mimeType", mimeType)
+            }
+            details.size?.let { size ->
+              putInt("size", size)
+            }
```
Comment (lukmccall) [inline packages/expo-document-picker/android/src/main/java/expo/modules/documentpicker/DocumentPickerModule.kt:80]: We can use a custom `Record` here. It'll be more readable.

File: packages/expo-document-picker/android/src/main/java/expo/modules/documentpicker/DocumentPickerModule.kt [hunk 3]
```diff
@@ -5,116 +5,93 @@ import android.content.Context
...
           }
-          documentDetails.size?.let {
-            putInt("size", documentDetails.size)
+          promise.resolve(result)
+        } ?: throw FailedToReadDocumentException()
+      } else {
+        promise.resolve(
+          Bundle().apply {
+            putString("type", "cancel")
           }
```
Comment (lukmccall) [inline packages/expo-document-picker/android/src/main/java/expo/modules/documentpicker/DocumentPickerModule.kt:88]: Same here

File: packages/expo-document-picker/android/src/main/java/expo/modules/documentpicker/DocumentPickerOptions.kt [hunk 4]
```diff
@@ -1,51 +1,9 @@
...
-
-  override fun hashCode(): Int {
-    var result = copyToCacheDirectory.hashCode()
-    result = 31 * result + types.contentHashCode()
-    return result
-  }
-}
+data class DocumentPickerOptions(
+  @Field val copyToCacheDirectory: Boolean,
+  @Field val type: List<String>
```
Comment (lukmccall) [inline packages/expo-document-picker/android/src/main/java/expo/modules/documentpicker/DocumentPickerOptions.kt:8]:
```suggestion
  @Field @IsNotEmpty val type: List<String>
```
To ensure that list isn't empty, you can use the `IsNotEmpty` annotation from https://github.com/expo/expo/blob/767aea9caa8cc49fdbf2be727b350fdceaca4358/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/records/Validators.kt#L46
The code will automatically check if the collection has at least one item.

**Review by lukmccall — CHANGES_REQUESTED** [review]
Looking promising ;)

File: packages/expo-document-picker/android/src/main/java/expo/modules/documentpicker/DocumentPickerOptions.kt (hunk 4 above)
Comment (alanjhughes) [reply packages/expo-document-picker/android/src/main/java/expo/modules/documentpicker/DocumentPickerOptions.kt]: Sorry, I had seen these 😅

**Review by lukmccall — APPROVED** [review]
🔥 🔥 🔥

**Pull request merged**